behavioral/chain.py

```diff
@@ -3,192 +3,95 @@
 
 """
 *What is this pattern about?
-This pattern aims to decouple the senders of a request from its
-receivers. It does this by allowing a request to move through chained
-objects until it is handled by an appropriate receiver.
 
-This is useful as it reduces the number of connections between objects,
-since the sender does not need explicit knowledge of the handler, and
-the receiver won't need to refer to all potential receivers, but keeps
-a reference to a single successor.
+The Chain of responsibility is an object oriented version of the
+`if ... elif ... elif ... else ...` idiom, with the
+benefit that the condition–action blocks can be dynamically rearranged
+and reconfigured at runtime.
+
+This pattern aims to decouple the senders of a request from its
+receivers by allowing request to move through chained
+receivers until it is handled.
 
-*References:
-http://www.dabeaz.com/coroutines/
+Request receiver in simple form keeps a reference to a single successor.
+As a variation some receivers may be capable of sending requests out
+in several directions, forming a `tree of responsibility`.
 
 *TL;DR80
-Allow a request to pass down a chain of objects until an object handles
-the request.
+Allow a request to pass down a chain of receivers until it is handled.
 """
 
-from contextlib import contextmanager
-import os
-import sys
-import time
 import abc
 
 
 class Handler(object):
     __metaclass__ = abc.ABCMeta
 
     def __init__(self, successor=None):
-        self._successor = successor
+        self.successor = successor
 
     def handle(self, request):
-        res = self._handle(request)
-        if not res:
-            self._successor.handle(request)
+        """
+        Handle request and stop.
+        If can't - call next handler in chain.
+
+        As an alternative you might even in case of success
+        call the next handler.
+        """
+        res = self.compare(request)
+        if not res and self.successor:
+            self.successor.handle(request)
 
     @abc.abstractmethod
-    def _handle(self, request):
-        raise NotImplementedError('Must provide implementation in subclass.')
+    def compare(self, request):
```

Comment:
I suggest to find a better name for `compare` and make it static (using @staticmethod). `compare` as a class method means compare this object/instance to another object/instance of the same class (e.g. implemented as `__cmp__(self, other)`

Reply:
Agree. Let's come up with a name. "check_range"? (naming is hard :)

Reply:
Yes,` check_range` or `check_request` might be better names :)